### PR TITLE
Separate Cache Lookup and Token Refresh in SilentFlowClient

### PR DESCRIPTION
**Pull request by tnorling**
This PR mainly refactors the SilentFlowClient and provides 2 additional APIs

Changes include:
- Addition of `acquireCachedToken` and `refreshToken` function in `SilentFlowClient`
- Original `acquireToken` function refactored to call the 2 new APIs
- Move the token cache lookups to `CacheManager`
- Add `getCacheRecord` function to `CacheManager` to do all the token cache lookups in one place

### CODE REVIEW

File: lib/msal-common/src/cache/entities/IdTokenEntity.ts [hunk 1]
```diff
@@ -53,6 +57,27 @@ export class IdTokenEntity extends CredentialEntity {
     }
 
     /**
+     * fetches idToken from cache if present
+     * @param request
+     */
+    static readIdTokenFromCache(cacheManager: CacheManager, clientId: string, account: AccountInfo, realm?: string): IdTokenEntity {
```
Comment (pkanher617) [inline lib/msal-common/src/cache/entities/IdTokenEntity.ts:63]: Maybe these should be helpers on CacheManager instead?

File: lib/msal-common/src/client/SilentFlowClient.ts [hunk 2]
```diff
@@ -35,6 +31,25 @@ export class SilentFlowClient extends BaseClient {
      * @param request
      */
     public async acquireToken(request: SilentFlowRequest): Promise<AuthenticationResult> {
+        let result: AuthenticationResult;
+        try {
+            result = this.acquireCachedToken(request);
```
Comment (jasonnutter) [inline lib/msal-common/src/client/SilentFlowClient.ts:36]: Only needs to be `return this.acquireCachedToken(request)`;

File: lib/msal-common/src/cache/entities/IdTokenEntity.ts [hunk 3]
```diff
@@ -53,6 +57,27 @@ export class IdTokenEntity extends CredentialEntity {
...
+        if (StringUtils.isEmpty(realm)) {
+            const accountKey: string = AccountEntity.generateAccountCacheKey(account);
+            const cachedAccount = cacheManager.getAccount(accountKey);
+            realm = cachedAccount && cachedAccount.realm ? cachedAccount.realm : realm;
```
Comment (jasonnutter) [inline lib/msal-common/src/cache/entities/IdTokenEntity.ts:67]: Would prefer to not reassign `realm` (even if its empty).

Comment (jasonnutter) [reply lib/msal-common/src/cache/entities/IdTokenEntity.ts:67]: The `isEmpty` check is only needed for the `realm` value, the preceding two lines don't need to be guarded by it.

Comment (pkanher617) [reply lib/msal-common/src/cache/entities/IdTokenEntity.ts:67]: the cached account is not needed if the realm is not empty. this saves a cache lookup.

Comment (jasonnutter) [reply lib/msal-common/src/cache/entities/IdTokenEntity.ts:67]: What if a helper function was used?

```js
const accountRealm = StringUtils.isEmpty(realm) ? this.getRealmForAccount(account) : realm;
```

File: lib/msal-common/src/cache/entities/IdTokenEntity.ts (hunk 1 above)
Comment (tnorling) [reply lib/msal-common/src/cache/entities/IdTokenEntity.ts:63]: Yeah I considered that as well. I initially thought this made more organizational sense but after thinking more on it I'm inclined to agree with you

File: lib/msal-common/src/client/SilentFlowClient.ts (hunk 2 above)
Comment (tnorling) [reply lib/msal-common/src/client/SilentFlowClient.ts]: Done.

File: lib/msal-common/src/cache/entities/IdTokenEntity.ts (hunk 1 above)
Comment (tnorling) [reply lib/msal-common/src/cache/entities/IdTokenEntity.ts]: Done.

File: lib/msal-common/src/cache/entities/IdTokenEntity.ts (hunk 3 above)
Comment (tnorling) [reply lib/msal-common/src/cache/entities/IdTokenEntity.ts]: Done.

File: lib/msal-common/src/cache/CacheManager.ts [hunk 4]
```diff
@@ -168,10 +186,71 @@ export abstract class CacheManager implements ICacheManager {
...
+     * @param inputRealm 
+     */
+    getAccessTokenEntity(clientId: string, account: AccountInfo, scopes: ScopeSet): AccessTokenEntity | null {
+        const accessTokenFilter: CredentialFilter = {
+            homeAccountId: account.homeAccountId,
+            environment: account.environment,
+            credentialType: CredentialType.ACCESS_TOKEN,
+            clientId,
+            realm: account.tenantId,
+            target: scopes.printScopesLowerCase()
```
Comment (sameerag) [inline lib/msal-common/src/cache/CacheManager.ts:225]: We are storing `scopes` in lower case and `look-up` in lower case. We need to address this, as I observed in cache-compat test cases, other libraries store the case as is. We may need to talk about this.

My take: Look up should be case insensitive but storing should be as is.

Comment (tnorling) [reply lib/msal-common/src/cache/CacheManager.ts:225]: 100% agree. As this PR is scoped to a refactor, let's take functional changes in a follow up PR

**Pull request merged**